libs/pika/async_mpi/src/mpi_polling.cpp

```diff
@@ -688,155 +693,182 @@ namespace pika::mpi::experimental {
-            flags &= ~1;
+            PIKA_DETAIL_DP(detail::mpi_debug<1>,
+                debug(str<>("register_pool"), "pool =", pool_name, ", mode",
+                    mode_string(get_completion_mode()), get_completion_mode()));
+            if (pool_name == resource::get_partitioner().get_default_pool_name())
+                detail::custom_pool_exists_ = false;
+            else
+            {
+                for (std::size_t p = 1; p < resource::get_partitioner().get_num_pools(); ++p)
+                {
+                    if (pool_name == resource::get_partitioner().get_pool_name(p))
+                        detail::custom_pool_exists_ = true;
+                }
+                if (!detail::custom_pool_exists_)
+                {
+                    PIKA_THROW_EXCEPTION(pika::error::bad_parameter, "mpi::register",
+                        "Register failed '{}' does not exist", pool_name);
+                }
+            }
+            detail::polling_pool_name_ = pool_name;
         }
-        PIKA_DETAIL_DP(detail::mpi_debug<1>,
-            debug(str<>("completion mode"), bin<8>(flags), detail::mode_string(flags)));
-        // override the variable used to control completion mode and pool flags
-        setenv("PIKA_MPI_COMPLETION_MODE", std::to_string(flags).c_str(), true);
 
-        // if pool is now disabled, just exit
-        if (!detail::use_pool(flags))
+        // -----------------------------------------------------------------
+        void register_polling()
         {
-            set_pool_name("default");
-            return false;
+            auto mode = get_completion_mode();
+            // enable pika polling on the mpi pool if the handling mode needs it
+            if (detail::get_handler_method(mode) != handler_method::yield_while)
+            {
+                PIKA_DETAIL_DP(detail::mpi_debug<1>,
+                    debug(str<>("enabling polling"), "pool =", get_pool_name(), ", mode",
+                        mode_string(get_completion_mode()), get_completion_mode()));
+                detail::register_polling(pika::resource::get_thread_pool(get_pool_name()));
+            }
         }
-        // Disable idle backoff on the MPI pool
-        using pika::threads::scheduler_mode;
-        auto smode = scheduler_mode::default_mode & ~scheduler_mode::enable_idle_backoff;
 
-        // Create a thread pool with a single core that we will use for all
-        // communication related tasks
-        std::string name = pool_name;
-        if (name.empty()) { name = get_pool_name(); }
-        else
+        // -----------------------------------------------------------------
+        void unregister_polling()
         {
-            // override mpi pool name with whatever we decided on
-            set_pool_name(name);
+            auto mode = get_completion_mode();
+            // enable pika polling on the mpi pool if the handling mode needs it
```

Comment:
The main issue is that we need to reserve a core before the main init is done, so adding a flag to init in place of the existing callback made sense, but I suppose I could simply add a direct call into pika::mpi to do the pool create during init and not have the user set a flag at all. 
Seems a bit intrusive to create pools without the user explicitly asking for one - likewise having polling enabled by default is ok if the pool is custom, but on the default pool - hmmm. Users might well compile pika with mpi support and use it for app1, but also app2 and might not use mpi much or at all and it's better to ask to enable - than ask to disable a feature like that.

I would agree that having 
`--pika:mpi:enable-pool` and `--pika:mpi:completion-mode=42` is the way to go - how about if I leave the pool init as it is for now and in a new pr, add command line option parsing on a module-by-module basis so that we can have explicit `pika:mpi:` options when mpi is enabled and not when it is not

Reply:
on second thoughts, I'll add a flag for `mpi-enable-pool` to complement `completion-mode` in this PR and think about module stuff later - NB the completion modes will need to change numbering once pool is removed from the bitmask.

Reply:
I'd prefer to remove modes in a separate PR

Reply:
Yeah, perfectly fine by me (I also prefer it that way). We should try to do it before the next release though.